### ISSUE TITLE
Experiment with message ordering

### ISSUE THREAD

**Issue by biobootloader** [AbanteAI/archive-old-cli-mentat#117]
Currently the conversation we send to the model is structured like this:

1. initial system prompt, describing the edit format that needs to be followed
2. user message
3. assistant response
4. user message
5. system "code message", containing all the context of the codebase (included files, diffs, codemaps, etc)

(in this example, the model already responded previously once, the structure is the same for shorter or longer conversations, always with a "code message" from the system as the final message)

This order is likely not optimal, because models seem to pay the most attention to the end of the prompt, which here we are filling with code context. My best guess for a better ordering would be:

1. system prompt describing edit format
2. system "code message"
3. user message
4. assistant message
5. user message

This would put the most important thing last (the user message). It may be better to reverse the order of (1) and (2). We should experiment to see if this improves performance at all. However running these experiments effectively will require longer context benchmarks than we currently have.

**Comment by biobootloader** [comment]
@jakethekoenig this could also be tied to what we discussed about sending previous assistant messages as system messages. Perhaps we'd just have the whole previous conversation in a single system message (and at some point, summarized)

**Comment by jakethekoenig** [comment]
Why would these experiments require longer context? Isn't it a matter of switching two messages which should have no effect on total length?

Another thing that concerns me: if the "code message" always has an up to date version of the code then if we put at the beginning it could be confusing to the model if the user is asked for a change that appears to be already present in the code. Summarizing the transcript could be quite helpful there. Instead of writing:
3. User Message
4. Assistant Message
we could write something like:
3. System: "The user previously requested {user request} for which the model changed {file list} and the user accepted.

**Comment by biobootloader** [comment]
> Why would these experiments require longer context? Isn't it a matter of switching two messages which should have no effect on total length?

You're right it won't change the total length, what I meant was that if we aren't using much context then the order probably doesn't matter, so we won't see a difference. But once we are using a lot of context it might matter a lot (would be even good to test with 32k model).

Like for Claude, Anthropic tested asking questions where the answers are in the beginning, middle, or end of long context prompts, and you can see big changes in accuracy (https://www.anthropic.com/index/prompting-long-context):

<img width="879" alt="image" src="https://github.com/AbanteAI/mentat/assets/128252497/fe2d7afc-e7b1-49ad-a124-717f390defda">

**Comment by biobootloader** [comment]
> Another thing that concerns me: if the "code message" always has an up to date version of the code then if we put at the beginning it could be confusing to the model if the user is asked for a change that appears to be already present in the code. Summarizing the transcript could be quite helpful there.

Good point, I like your idea!
